### PR TITLE
Implement clone method

### PR DESCRIPTION
**Pull request by talyssonoc**

### CODE REVIEW

File: dist/structure.js [hunk 1]
```diff
@@ -1329,6 +1333,35 @@ return /******/ (function(modules) { // webpackBootstrap
...
+/***/ },
+/* 39 */
+/***/ function(module, exports) {
+
+	"use strict";
+
+	exports.buildCloneDescriptorFor = function buildCloneDescriptorFor(StructureClass) {
+	  return {
+	    value: function clone() {
+	      var overwrites = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
```
Comment (negreirosleo) [inline dist/structure.js]: I think it's a good idea do use let here

File: docs/cloning.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+Structure adds a method `#clone` in order to be able to create a **shallow** copy of an instance. This methods accepts an optional overwrite object that permits you to overwrite some attributes of the copy.
+
+```js
+const { attributes } = require('structure');
+
+const User = attributes({
+  name: String,
+})(class User {});
+
+var user = new User({
```
Comment (negreirosleo) [inline docs/cloning.md]: I think it's a good idea to use let here [2]

File: docs/cloning.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+const Book = attributes({
+  name: String,
+})(class Book {});
+
+const User = attributes({
+  name: String,
+  favoriteBook: Book,
+})(class User {});
+
+var user = new User({
```
Comment (negreirosleo) [inline docs/cloning.md]: I think it's a good idea to use let here [3]

File: docs/cloning.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    required: true,
+  },
+  age: Number,
+})(class User {});
+
+const user = new User({
+  name: 'Me',
+});
+
+var clonedUser = user.clone(
```
Comment (negreirosleo) [inline docs/cloning.md]: I think it's a good idea to use let here [4]

File: test/unit/instanceAndUpdate.spec.js [hunk 5]
```diff
@@ -369,3 +369,210 @@ describe('updating a structure with dynamic attribute types', () => {
     expect(user.favoriteBook.owner).to.equal(user);
   });
 });
+
+describe('cloning an instance', () => {
+  var User;
+  var Book;
```
Comment (negreirosleo) [inline test/unit/instanceAndUpdate.spec.js]: I think it's a good idea to use let here [5]

File: dist/structure.js [hunk 6]
```diff
@@ -1329,6 +1333,35 @@ return /******/ (function(modules) { // webpackBootstrap
...
+	      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
+	      var strict = options.strict;
+
+
+	      var newAttributes = Object.assign({}, this.attributes, overwrites);
+
+	      var cloneInstance = void 0;
```
Comment (negreirosleo) [inline dist/structure.js]: I think it's a good idea to use let here

**Comment by talyssonoc** [comment]
@loweoj you can do it explicitly:

```js
const clonedUser = user.clone({ favoriteBook: user.favoriteBook.clone() })
```

I want to leave it open that way for cases when you want to clone some of the nested structures but not others, this is the most general case.

**Pull request merged**